src/server/requestHandlers.js

```diff
@@ -222,42 +188,85 @@ sjrk.storyTelling.server.saveStoryToDatabase = function (dataSource, binaryRenam
+            funcName: "sjrk.storyTelling.server.handleSaveStoryFile",
+            args: ["{arguments}.0", "{server}.storyDataSource", "{server}.options.globalConfig.authoringEnabled"]
+        }
+    }
+});
+
 /**
- * Update any media URLs to refer to the changed file names
+ * Saves a single file to the server filesystem and responds to the HTTP request.
+ * If the file is an image, it will be rotated to match its EXIF orientation data.
  *
- * @param {Object} blocks - a collection of story blocks with file references
- * @param {Object} files - a list of files to refer to
+ * Errors will be raised in the following situations:
+ * - Authoring is not enabled
+ * - An error is encountered while trying to get the story (invalid ID, DB offline, etc.)
+ * - The story that the file is associated with is already published
+ * - The provided file is not valid
+ * - An error is encountered while rotating an image to its correct orientation
  *
- * @return {Object.<String, String>} - a key-value collection linking uploaded filenames to server paths
+ * @param {Object} request - a Kettle request containing a single file associated with a story
+ * @param {Component} dataSource - an instance of sjrk.storyTelling.server.dataSource.couch.story
+ * @param {Boolean} authoringEnabled - a server-level flag to indicate whether authoring is enabled
  */
-sjrk.storyTelling.server.buildBinaryRenameMap = function (blocks, files) {
-    // key-value pairs of original filename : generated filename
-    // this is used primarily by tests, but may be of use
-    // to client-side components too
-    var binaryRenameMap = {};
-
-    fluid.each(blocks, function (block) {
-        if (block.blockType === "image" || block.blockType === "audio" || block.blockType === "video") {
-            if (block.fileDetails) {
-                // Look for the uploaded file matching this block
-                var mediaFile = fluid.find_if(files, function (file) {
-                    return file.originalname === block.fileDetails.name;
-                });
+sjrk.storyTelling.server.handleSaveStoryFile = function (request, dataSource, authoringEnabled) {
+    if (!authoringEnabled) {
+        request.events.onError.fire({
+            isError: true,
+            message: "Saving is currently disabled."
+        });
+
+        return;
+    }
 
-                // If we find a match, update the media URL. If not, clear it.
-                if (mediaFile) {
-                    sjrk.storyTelling.server.setMediaBlockUrl(block, mediaFile.filename);
-                    binaryRenameMap[mediaFile.originalname] = mediaFile.filename;
-                } else {
-                    sjrk.storyTelling.server.setMediaBlockUrl(block, null);
-                }
+    var id = request.req.params.id;
+
+    // verify that the story exists and isn't published before continuing
+    dataSource.get({directStoryId: id}).then(function (story) {
+        if (story && !story.published && request.req.file) {
```

Comment:
Is it possible for the `dataSource.get` to resolve without a response?

Reply:
I believe this won't ever be the case, so the check for `story` is probably overkill. Removing it for now. If we discover a case where it *is* possible, then we'll want to add the same check to `sjrk.storyTelling.server.handleGetStory`